### ISSUE TITLE
Performance degrades after ~2 days of uptime, setting sync=disabled mitigates the issue

### ISSUE THREAD

**Issue by fiesh** [openzfs/zfs#16658]
### System information

Type | Version/Name
 --- | ---
Distribution Name	| Ubuntu
Distribution Version	| 24.04
Kernel Version	| 6.8.0-45-generic
Architecture	| amd64
OpenZFS Version	| 2.2.2

### Describe the problem you're observing

Performance seriously degrades after about 48h of uptime.  It becomes very noticeable that operations requiring a sync become very slow.  This is particularly noticeable via NFS.

### Describe how to reproduce the problem

This appears just after sufficient uptime.  The system works fine upon a fresh reboot.  Setting `sync=disabled` makes the performance regression go away immediately, switching back to `sync=standard` makes it re-appear.  It feels like every sync call takes about one second to complete.

### Include any warning/errors/backtraces from the system logs

I would be very thankful for tips on how to debug this.  I couldn't find anything of interest in any logging.

**Comment by amotin** [comment]
There is not enough information here to say anything.  Where do you write your ZIL: normal vdevs or SLOG?  Is that device designed for ZIL usage?  Have you measured write/sync latency of that device as your performance degrade?

**Comment by fiesh** [comment]
Thanks for your answer.  This is the pool in question, the devices are all brand new:

```
  pool: rpool
 state: ONLINE
  scan: scrub repaired 4K in 00:41:19 with 0 errors on Sun Oct 13 01:05:21 2024
config:

        NAME                                                                  STATE     READ WRITE CKSUM
        rpool                                                                 ONLINE       0     0     0
          raidz2-0                                                            ONLINE       0     0     0
            nvme-Samsung_SSD_990_PRO_with_Heatsink_2TB_S766NJ0W501783L-part4  ONLINE       0     0     0
            nvme-Samsung_SSD_990_PRO_with_Heatsink_2TB_S766NJ0W501778A-part4  ONLINE       0     0     0
            nvme-Samsung_SSD_990_PRO_with_Heatsink_2TB_S766NJ0W502072J-part4  ONLINE       0     0     0
            nvme-Samsung_SSD_990_PRO_with_Heatsink_2TB_S766NJ0W501781D-part4  ONLINE       0     0     0
            nvme-Samsung_SSD_990_PRO_with_Heatsink_2TB_S766NJ0W501779B-part4  ONLINE       0     0     0
            nvme-Samsung_SSD_990_PRO_with_Heatsink_2TB_S766NJ0W501787K-part4  ONLINE       0     0     0
            nvme-Samsung_SSD_990_PRO_with_Heatsink_2TB_S766NJ0W501785E-part4  ONLINE       0     0     0
            nvme-Samsung_SSD_990_PRO_with_Heatsink_2TB_S766NJ0W501773P-part4  ONLINE       0     0     0

errors: No known data errors
```

I haven't measured the latency and would be surprised if the devices performed poorly, but of course I can do that. How should I measure it?

I was quite baffled by the scrub having reported a checksum error on one of the devices.  I ran a full self-test on the device via the `nvme` cli and it responded no issues.  I cleared the zpool error and nothing has surfaced since.

**Comment by amotin** [comment]
Those drives are desktop-oriented TLC.  It means that after writing about 10% of their capacity at high speed their performance is expected to drop significantly unless they are given enough time to rest and recover internally.  I don't know how much have you written there in the 48 hours, but it is one of the options.

On top of that, wide RAIDZ2 of desktop SSDs is not great as SLOG, since it makes each SSD to commit data to flash, that might be expensive for SSDs without power loss protection (AKA big capacitor on the board).  For such transaction-heavy workload with a lot of syncs I would probably reconfigure the pool as 2x 4-wide RAIDZ1.  Alternatively you could add separate SLOG device with even faster write-oriented power-loss-protected SSD or NVMe.  And even more alternatively you may think whether you need protection for crashes or sync=disabled might be fine for your workload.

**Comment by fiesh** [comment]
Thank you very much for the primer on SSDs / NVMes, I wasn't aware of that.  The devices have, over their lifetime, 1.89 TB of data units written, which doesn't sound like a lot to me but of course is far above the 10 % threshold you mentioned.

The system is protected against power loss, and I already have `sync=disabled` set for datasets with lots of ephemeral writing.  So I hoped to have it set to `standard` for more static datasets.

Is there some way for me to verify that it is indeed the device whose synchronization locks for such a long time?  If so, this would of course be an incorrect issue that I'd have to close.  Double-checking would be great.

**Comment by fiesh** [comment]
Closing this for now since it's probably not ZFS related but something fishy underneath.  Thank you for your help, @amotin.

**Comment by fiesh** [comment]
So to come back to this, I believe we found the issue, and it was not ZFS related.  I'll add some explanation though for any unfortunate person who also runs into this:

The Samsung SSD 990 Pro devices that we had came with a firmware version that randomly triggers devices to report a temperature of 84 degrees Celsius and never go back from that.  This then seems to lead to some slowed down mode to keep the device from completely overheating, thus slowing down the entire pool.  Us having 8 of these devices probably exacerbated the issue by increasing the likelihood of this happening.

We have now updated to firmware version `4B2QJXD7`, which, according to Samsung's website _To address reports of high temperatures logged on Samsung Magician._

As a side note: Samsung's un-bootable UEFI memory image contains an update utility aptly called `fumagician` that can be used directly from a working Linux installation.  There is no need to follow the screenshot-filled, entirely broken instruction manual written by morons requiring GUI tools to get the image to boot.  Simply do a `gzip -dc /path/to/initrd | cpio -idv --no-absolute-filenames` on the mounted image to extract the tool that Samsung chooses to keep obscurely hidden.
